### PR TITLE
feat: 'add' command accepts directories

### PR DESCRIPTION
**Pull request by pyramation**
## added missing feature for git add

I'm not sure if there are any more edge cases you'd like to cover, but wanted to kick off the conversation with a PR. Happy to make more fixes to make this mergeable. Cheers!

### CODE REVIEW

**Comment by karma-pr-reporter** [comment]
## Test Results for 8fdf00d7b943bbd9e8b478ff3b1cefa2e176a0a3:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 226 | 1 | 0 | 58.665 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 226 | 1 | 0 | 1 min 36.824 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 226 | 1 | 0 | 2 mins 11.228 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 226 | 1 | 0 | 3 mins 8.662 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 226 | 1 | 0 | 3 mins 19.051 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 226 | 1 | 0 | 4 mins 18.177 secs | false | undefined |

**Comment by pyramation** [comment]
I just found a corner case. When you try and emulate `git add .` it will add the `gitdir`. Working on this case, but, I'm assuming there would also be a `.gitignore` which can add complexity since it can exist in any folder in a repo.

File: src/commands/add.js [hunk 1]
```diff
@@ -23,7 +23,21 @@ export async function add ({
     let stats = await fs.lstat(join(dir, filepath))
     if (!stats) throw new GitError(E.FileReadError, { filepath })
     if (stats.isDirectory()) {
-      throw new GitError(E.NotImplementedFail)
+      const children = await fs.readdir(join(dir, filepath))
+      const gitdirRegExp = new RegExp(gitdir);
+      for (var c = 0; c < children.length; c++) {
+        if (gitdirRegExp.test( join(dir, filepath, children[c]) )) {
```
Comment (wmhilton) [inline src/commands/add.js]: I think this should probably just be `if (children[c] === '.git') {`

**Comment by wmhilton** [comment]
Yes, we should probably integrate `GitIgnoreManager` into it, because otherwise people will be quite shocked when it adds their entire `node_modules` folder and they fall into a black hole.

**Comment by karma-pr-reporter** [comment]
## Test Results for b5a1b4aa9a5d930fb7edc6e2b8dc89dfc5f88f85:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 229 | 1 | 0 | 59.581 secs | false | undefined |
| Firefox 63.0.0 (Ubuntu 0.0.0) | 229 | 1 | 0 | 1 min 41.389 secs | false | undefined |
| Safari 12.0.0 (Mac OS X 10.13.6) | 229 | 1 | 0 | 7 mins 36.068 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 229 | 1 | 0 | 7 mins 56.827 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 229 | 1 | 0 | 8 mins 36.955 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 229 | 1 | 0 | 8 mins 51.395 secs | false | undefined |

**Comment by pyramation** [comment]
Yea, I was already thinking that step one is correctness, and now we can look at speed. One thing that comes to mind is the caching of the `GitIgnoreManager`, since *for every file* it's reading all the directories. I think if we can cache the file reads that the manager is doing, it should help quite a bit

**Comment by wmhilton** [comment]
Currently it's reading every single file. If we used `status` and only `add` files that aren't already added that should help... but I tried that and it only shaved off 10%. So I think we'll need to dive deep and profile the sucker to find out where all this time is getting spent. (I also might need to read some more git source code and see what optimizations they do.)

**Comment by wmhilton** [comment]
You might be onto something with the GitIgnoreManager & cacheing. For every file, it at least runs `fs.stat` for every sub folder. If that's the bottleneck that would explain why running `status` didn't help.

**Comment by pyramation** [comment]
another thought I had was around the implementation of `writeObject` - which emulates the `write-tree`. If it is calculating the folders' `tree` for each time you `add`, one idea is to make a private version that only writes blobs from worktree to index, and then does the calculation of the entire tree after all additions to index happen.

**Comment by wmhilton** [comment]
@pyramation `git add` never writes trees. The only times trees are added to the index is during `git commit`. The current implementation here of `add` never calls `writeObject` with a directory. So that's not a contributing factor to speed.

**Comment by wmhilton** [comment]
Good news! I found a trivial change that makes it 5x faster.

Bad news: I feel compelled to create a screencast / tutorial about how to profile node applications and identify slow code paths, so I won't get to it until this weekend. I feel like it is a skill that a lot of devs don't have, but would definitely benefit from. Particularly anyone interested in contributing to `isomorphic-git` where speed makes a huge difference in the user experience and overall value.

**Comment by pyramation** [comment]
@wmhilton sounds great! WOW. That's huge. I'm super curious now--using flame charts?

**Comment by wmhilton** [comment]
Yes! Using flame charts 🔥 📊

**Comment by wmhilton** [comment]
Oh dear. How do people make excellent short coding instructional videos? I thought I was well prepared, but I still have a freaking 35 minutes of raw footage to edit now, bahaha. It'll take longer to make this video than to make the code 1000x faster. 😆

**Comment by karma-pr-reporter** [comment]
## Test Results for 28ea3240157230012e016ab6fcfc291dca05dd1b:
| Browser | Passed | Skipped | Failed | Time | Disconnected | Error |
|---------|--------|---------|--------|------|--------------|-------|
| HeadlessChrome 0.0.0 (Linux 0.0.0) | 246 | 1 | 0 | 1 min 15.069 secs | false | undefined |
| Firefox 64.0.0 (Ubuntu 0.0.0) | 246 | 1 | 0 | 2 mins 27.539 secs | false | undefined |
| Safari 11.1.2 (Mac OS X 10.13.6) | 246 | 1 | 0 | 3 mins 15.635 secs | false | undefined |
| Chrome Mobile 67.0.3396 (Android 7.1.1) | 246 | 1 | 0 | 3 mins 41.179 secs | false | undefined |
| Mobile Safari 11.0.0 (iOS 11.2.0) | 246 | 1 | 0 | 4 mins 34.024 secs | false | undefined |
| Edge 17.17134.0 (Windows 10 0.0.0) | 246 | 1 | 0 | 5 mins 14.99 secs | false | undefined |

**Comment by wmhilton** [comment]
I still think we could make this 10x faster.... but I also really want to use this feature in a project I'm working on today. :-) So... need trumps speed lol.
